File: kotlin/react/src/main/kotlin/React.kt
class Reactor<T : Any> {

    abstract inner class Cell() {
        abstract val value: T
        private val listeners = mutableListOf<ComputeCell>()

        internal fun addListener(cell: ComputeCell) {
            listeners.add(cell)
        }

        internal fun recomputeListeners() {
            listeners.forEach { it.compute() }
        }

        internal fun fireListenerCallbacks() {
            listeners.forEach { it.fireCallbacks() }
        }
    }

    /* ************************************************** */
    inner class InputCell(data: T) : Cell() {
        override var value = data
            set(newData) {
                field = newData
                recomputeListeners()
                fireListenerCallbacks()
            }
    }

    /* ************************************************** */
    inner class ComputeCell(private vararg var cells: Cell, private val transform: (List<T>) -> T) : Cell() {

        override lateinit var value: T    // type T cannot be nullable for lateinit
            private set

        init {
            compute()
            cells.forEach { it.addListener(this) }
        }

        private var previousValue = value
        private val callbacks = mutableListOf<Callback>()

        internal fun compute() {
            value = transform(cells.map { it.value })
            recomputeListeners()
        }

        fun addCallback(func: (T) -> Unit): Subscription {
            val callback = Callback(this, func)
            callbacks.add(callback)
            return callback
        }

        internal fun removeCallback(callback: Callback) {
            callbacks.remove(callback)
        }

        internal fun fireCallbacks() {
            if (value != previousValue) {
                previousValue = value
                callbacks.forEach { it.fire() }
                fireListenerCallbacks()
            }
        }
    }

    /* ************************************************** */
    interface Subscription {
        fun cancel()
    }

    inner class Callback(private val cell: ComputeCell, val callback: (T) -> Unit): Subscription {
        internal fun fire() {
            callback(cell.value)
        }

        override fun cancel() {
            cell.removeCallback(this)
        }
    }
}

File: nim/react/react.nim
import std/sequtils
 
type
  Callback = proc(val: int)
  ComputeFunc = proc(vals: seq[int]): int {.noSideEffect.}  # func "type"

  Cell* = ref object of RootObj
    value: int
    listeners: seq[ComputeCell]
    callbacks: seq[Callback]

  InputCell = ref object of Cell

  ComputeCell = ref object of Cell
    inputs: seq[Cell]
    previousValue: int
    computeFunction: ComputeFunc

############################################################
# Cell procs
#
# a couple of forward declarations
proc recompute(cell: ComputeCell)
proc fireCallbacks(cell: ComputeCell)

proc recomputeListeners(cell: Cell) =
  for c in cell.listeners:
    c.recompute()

proc fireListenerCallbacks(cell: Cell) =
  for c in cell.listeners:
    c.fireCallbacks()

proc value*(cell: Cell): int = cell.value

proc `value=`*(cell: Cell, value: int) =
  cell.value = value
  cell.recomputeListeners()
  cell.fireListenerCallbacks()

############################################################
# InputCell procs
#
proc newInputCell*(value: int): Cell =
  result = InputCell(value: value)

############################################################
# ComputeCell procs
#
proc newComputeCell*(inputs: seq[Cell], function: ComputeFunc): Cell =
  let computeCell = ComputeCell(inputs: inputs, computeFunction: function)
  for cell in inputs:
    cell.listeners.add computeCell
  computeCell.recompute()
  computeCell.previousValue = computeCell.value
  computeCell

proc recompute(cell: ComputeCell) =
  cell.value = cell.computeFunction(cell.inputs.mapIt(it.value))
  cell.recomputeListeners()

proc addCallback*(cell: Cell, callback: Callback) =
  cell.callbacks.add callback

proc fireCallbacks(cell: ComputeCell) =
  if cell.value != cell.previousValue:
    cell.previousValue = cell.value
    for cb in cell.callbacks:
      cb(cell.value)
    cell.fireListenerCallbacks()

proc removeCallback*(cell: Cell, callback: Callback) =
  cell.callbacks.keepItIf(it != callback)
